src/fstext/determinize-star-inl.h

```diff
@@ -468,99 +469,221 @@ template<class Arc> class DeterminizerStar {
+              std::cerr << "First string: ";
+              for (size_t i = 0; i < tmp_seq.size(); i++)
+                std::cerr << tmp_seq[i] << " ";
+              std::cerr << "\nSecond string: ";
+              repository_.SeqOfId(next_elem.string, &tmp_seq);
+              for (size_t i = 0; i < tmp_seq.size(); i++)
+                std::cerr << tmp_seq[i] << " ";
+              std::cerr << "\n";
+            }
+            throw std::runtime_error("Non-functional FST: cannot determinize.\n");
+          }
+          if (iter->second.in_queue) {
+            // since in queue, no check since we're propagating them anyway
+            iter->second.weight_to_process =
+                Plus(iter->second.weight_to_process, next_unprocessed_weight);
+            // probably should take it out of the queue, update the depth
+            // and put it back.
+            // but it seems it's fast enough already ;-)
+          } else {
+            // is in the subset but not in queue
+            iter->second.weight_to_process = Plus(iter->second.weight_to_process,
+                                                  next_unprocessed_weight);
+
+            // this is because the code in "else" below: the 
+            // iter->second.weight_to_process might not be Zero()
+            Weight weight = Plus(iter->second.element.weight,
+                                 iter->second.weight_to_process);
+
+            // What is done below is, we propagate the weight (by adding them
+            // to the queue only when the change is big enough;
+            // otherwise we just store the weight, until before returning
+            // we add the element.weight and weight_to_process together
+            iter->second.depth = max(iter->second.depth, this_depth + 1);
+            if (! ApproxEqual(weight, iter->second.element.weight, delta_)) {
+              // add extra part of weight to queue.
+              iter->second.in_queue = true;
+              queue.push(IdAndDepth(next_elem.state, iter->second.depth));
+            } else {
+              // we update the weight_to_process, but not add it to the queue
+              iter->second.in_queue = false;
+            }
+          }
+        }
+      }
+    }
+  }
+
   void EpsilonClosure(const vector<Element> &input_subset,
                       vector<Element> *output_subset) {
     // input_subset must have only one example of each StateId.
 
-    std::map<InputStateId, Element> cur_subset;
-    typedef typename std::map<InputStateId, Element>::iterator MapIter;
+    std::map<InputStateId, EpsilonClosureInfo> cur_subset;
+
+    typedef typename std::map<InputStateId, EpsilonClosureInfo>::iterator MapIter;
     {
       MapIter iter = cur_subset.end();
       for (size_t i = 0; i < input_subset.size(); i++) {
-        std::pair<const InputStateId, Element> pr(input_subset[i].state, input_subset[i]);
+        queue.push(IdAndDepth(input_subset[i].state, 0));
```

Comment:
Thanks- let us know how this affects the speed.
It should be a big effect.
Dan


On Thu, Oct 8, 2015 at 9:36 PM, hainan-xv <notifications@github.com> wrote:

> In src/fstext/determinize-star-inl.h
> <https://github.com/kaldi-asr/kaldi/pull/208#discussion_r41589978>:
>
> >      {
> >        MapIter iter = cur_subset.end();
> >        for (size_t i = 0; i < input_subset.size(); i++) {
> > -        std::pair<const InputStateId, Element> pr(input_subset[i].state, input_subset[i]);
> > +        queue.push(IdAndDepth(input_subset[i].state, 0));
>
> Done.
> … <#1504a3e373cbfcd4_>
> On Thu, Oct 8, 2015 at 9:18 PM, Daniel Povey <notifications@github.com>
> wrote: In src/fstext/determinize-star-inl.h <
> https://github.com/kaldi-asr/kaldi/pull/208#discussion_r41589135>: > { >
> MapIter iter = cur_subset.end(); > for (size_t i = 0; i <
> input_subset.size(); i++) { > - std::pair<const InputStateId, Element>
> pr(input_subset[i].state, input_subset[i]); > +
> queue.push(IdAndDepth(input_subset[i].state, 0)); I think you can avoid
> actually adding these things to the queue here-- which is probably where
> most of the time is being taken. Why don't you remove the 'queue.push'
> statement from this loop, and then after this loop, have another loop from
> 0 to input_subset.size(), where you set InputStateId id =
> input_subset[i].state int this_depth = 0; This will process things in the
> same order as your queue, because you'll be processing all the 0-depth
> things first anyway. — Reply to this email directly or view it on GitHub <
> https://github.com/kaldi-asr/kaldi/pull/208/files#r41589135>.
> -- - Hainan
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/208/files#r41589978>.
>